### ISSUE TITLE
Implement decorators

### ISSUE THREAD

**Issue by fwoelffel** [fwoelffel/nest-bull#53]
We should have some decorator to ease the module usability.

Here's a sample class using such decorators:

```typescript
@QueueListener(options?: { name?: string })
export class MyQueueListener {

  @Process(options?: { name?: string, concurrency?: number })
  processor(job: Job) {}

  @OnError()
  onErrorListener(error: Error): void {}
  
  @OnWaiting()
  onWaitingListener(jobId: JobId): void {}
  
  @OnActive()
  onActiveListener(job: Job, jobPromise: JobPromise): void {}
  
  @OnStalled()
  onStalledListener(job: Job): void {}
  
  @OnProgress()
  onProgressListener(job: Job, progress: any): void {}
  
  @OnCompleted()
  onCompletedListener(job: Job, result: any): void {}
  
  @OnFailed()
  onFailedListener(job: Job, error: Error): void {}
  
  @OnPaused()
  onPausedListener(): void {}
  
  @OnResumed()
  onResumedListener(job: Job): void {}
  
  @OnCleaned()
  onCleanedListener(jobs: Job[], status: JobStatusClean): void {}
  
  @OnDrained()
  onDrainedListener(): void {}
  
  @OnRemoved()
  onRemovedListener(job: Job): void {}

  @OnGlobalError()
  onGlobalErrorListener(error: Error): void {}
  
  @OnGlobalWaiting()
  onGlobalWaitingListener(jobId: JobId): void {}
  
  @OnGlobalActive()
  onGlobalActiveListener(job: Job, jobPromise: JobPromise): void {}
  
  @OnGlobalStalled()
  onGlobalStalledListener(job: Job): void {}
  
  @OnGlobalProgress()
  onGlobalProgressListener(job: Job, progress: any): void {}
  
  @OnGlobalCompleted()
  onGlobalCompletedListener(job: Job, result: Error): void {}
  
  @OnGlobalFailed()
  onGlobalFailedListener(job: Job, error: Error): void {}
  
  @OnGlobalPaused()
  onGlobalPausedListener(): void {}
  
  @OnGlobalResumed()
  onGlobalResumedListener(job: Job): void {}
  
  @OnGlobalCleaned()
  onGlobalCleanedListener(jobs: Job[], status: JobStatusClean): void {}
  
  @OnGlobalDrained()
  onGlobalDrainedListener(): void {}
  
  @OnGlobalRemoved()
  onGlobalRemovedListener(job: Job): void {}
  
}
```

I think this improvement will be greatly appreciated and will resolve some current issues/PR.

**Comment by dhritzkiv** [comment]
One thing to consider in design: the ability to pass functions or property names to decorators.

So something like this:

```ts
@QueueListener()
export class MyQueueListener {
   processorName() {
      return "foo";
   }

  @Process({
    name() {
      return this.processorName()
    }
  })
  processor(job: Job) {
  
  }
}
```

But better, obviously

**Comment by fwoelffel** [comment]
I'll be considering the ability to pass functions to the decorators.

**Comment by fwoelffel** [comment]
Closed by #56
